src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+// @flow
+
+import { traverseAst } from "./utils/ast";
+import { getSource } from "./sources";
+
+import type { Source, SourceId, Location, SourceScope } from "debugger-html";
+import type { NodePath, Node, Location as BabelLocation } from "babel-traverse";
+
+export type ScopeDeclaration = {
+  type: string,
+  start: Location,
+  end: Location,
+  bindings: any
+};
+
+export type ScopeDeclarations = ScopeDeclaration[];
+
+type GeneratedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: any,
+  children: ?(GeneratedScope[])
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  scopes: Array<TempScope>,
+  loc: BabelLocation,
+  names: any
+};
+
+function _createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  let result = {
+    type,
+    parent,
+    scopes: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.scopes.push(result);
+  }
+  return result;
+}
+
+function isNode(node?: Node, type: string): boolean {
+  return node ? node.type === type : false;
+}
+
+function _getFunctionScope(scope: TempScope): TempScope {
+  let s = scope;
+  while (s.type !== "Function" && s.type !== "Script") {
+    if (!s.parent) {
+      return s;
+    }
+    s = s.parent;
+  }
+  return s;
+}
+
+function _fromBabelLocation(
+  location: BabelLocation,
+  sourceId: SourceId
+): Location {
+  return {
+    sourceId,
+    line: location.line,
+    column: location.column
+  };
+}
+
+function _parseDeclarator(
+  declaratorId: Node,
+  targetScope: TempScope,
+  kind: string
+) {
+  if (isNode(declaratorId, "Identifier")) {
+    targetScope.names[declaratorId.name] = { kind, refs: [] };
+    return;
```

Comment:
we don't strictly need this return, since the rest of the body is else statements. on the other hand we can remove else and just use if statements with returns